CompilerSource/backend/GameData.cpp

```diff
@@ -8,19 +8,199 @@
 
 #include "GameData.h"
 
+#include "lodepng.h"
+
 #include <map>
 #include <string>
 #include <iostream>
 
+#include <zlib.h>
+
 using std::cout;
 using std::cerr;
 using std::endl;
 
+using TypeCase = buffers::TreeNode::TypeCase;
+
 static inline uint32_t javaColor(int c) {
   if (!(c & 0xFF)) return 0xFFFFFFFF;
   return ((c & 0xFF0000) >> 8) | ((c & 0xFF00) << 8) | ((c & 0xFF000000) >> 24);
 }
 
+static inline std::string string_replace_all(std::string str, std::string substr, std::string nstr)
+{
+  size_t pos = 0;
+  while ((pos = str.find(substr, pos)) != std::string::npos)
+  {
+    str.replace(pos, substr.length(), nstr);
+    pos += nstr.length();
+  }
+  return str;
+}
+
+static unsigned char* zlib_compress(unsigned char* inbuffer, int &actualsize)
+{
+  uLongf outsize=(int)(actualsize*1.1)+12;
+  Bytef* outbytef=new Bytef[outsize];
+
+  compress(outbytef,&outsize,(Bytef*)inbuffer,actualsize);
+
+  actualsize = outsize;
+
+  return (unsigned char*)outbytef;
+}
+
+static std::string Argument2Code(const buffers::resources::Argument& arg) {
+  using buffers::resources::ArgumentKind;
+  std::string val = arg.string();
+
+  switch (arg.kind()) {
+    case ArgumentKind::ARG_BOTH:
+      // treat as literal if starts with quote (")
+      if (val[0] == '"' || val[0] == '\'') return val;
+      // else fall through
+    case ArgumentKind::ARG_STRING:
+      return '\"' + string_replace_all(string_replace_all(val, "\\", "\\\\"), "\"", "\"+'\"'+\"") + '\"';
+    case ArgumentKind::ARG_BOOLEAN:
+      return std::to_string(val != "0");
+    case ArgumentKind::ARG_SPRITE:
+      return arg.has_sprite() ? arg.sprite() : "-1";
+    case ArgumentKind::ARG_SOUND:
+      return arg.has_sound() ? arg.sound() : "-1";
+    case ArgumentKind::ARG_BACKGROUND:
+      return arg.has_background() ? arg.background() : "-1";
+    case ArgumentKind::ARG_PATH:
+      return arg.has_path() ? arg.path() : "-1";
+    case ArgumentKind::ARG_SCRIPT:
+      return arg.has_script() ? arg.script() : "-1";
+    case ArgumentKind::ARG_OBJECT:
+      return arg.has_object() ? arg.object() : "-1";
+    case ArgumentKind::ARG_ROOM:
+      return arg.has_room() ? arg.room() : "-1";
+    case ArgumentKind::ARG_FONT:
+      return arg.has_font() ? arg.font() : "-1";
+    case ArgumentKind::ARG_TIMELINE:
+      return arg.has_timeline() ? arg.timeline() : "-1";
+    case ArgumentKind::ARG_MENU:
+    case ArgumentKind::ARG_COLOR:
+    default:
+      if (val.empty()) {
+        if (arg.kind() == ArgumentKind::ARG_STRING)
+          return "\"\"";
+        else
+          return "0";
+      }
+      // else fall all the way through
+  }
+
+  return val;
+}
+
+static std::string Actions2Code(const ::google::protobuf::RepeatedPtrField< buffers::resources::Action >& actions) {
```

Comment:
This is probably better off in a separate source. It's a hack from the LGM days that we may as well maintain, but ideally not in the game storage logic. I'd just move this whole block of additions to an `actions2code.cpp`/`h`.

Reply:
I moved it exactly the way you describe in #1479 (which I now have to redo after we get this finalized and merged) but just didn't want to over-complicate what I was doing here.

Reply:
Ok: 3c7c587f2702179ea0197e81bac61e7b17ae166b

I split it out on its own again.